jobs/job_test_utils.py

```diff
@@ -45,14 +49,32 @@ def __init__(self, *args, **kwargs):
             runner=runners.DirectRunner(),
             options=test_pipeline.PipelineOptions(runtime_type_check=True))
 
-    def run(self, *args, **kwargs):
-        """Runs the test within the context of a test pipeline."""
-        with self.pipeline:
-            super(PipelinedTestBase, self).run(*args, **kwargs)
+    def setUp(self):
+        super(PipelinedTestBase, self).setUp()
+        with contextlib2.ExitStack() as stack:
+            stack.enter_context(decorate_beam_errors())
+            stack.enter_context(self.pipeline)
+            self._close_stack = stack.pop_all().close
+
+    def tearDown(self):
+        """Hook method for deconstructing the test fixture after testing it.
+
+        Raises:
+            AssertionError. The pipeline wrapping the test fixture has failed.
+        """
+        # NOTE: Important to call super().tearDown() first so that all classes
+        # can clean up before trying to close the pipeline, which may raise.
+        # Normally tearDown() should call super() after it has finished its job.
+        super(PipelinedTestBase, self).tearDown()
+        # Now it's safe to try and close the pipeline.
+        self._close_stack()
 
     def assert_pcoll_equal(self, actual, expected):
         """Asserts that the given PCollections are equal.
 
+        NOTE: This method doesn't raise the exception. It attaches a PTransform
```

Comment:
Ooh -- interesting. Maybe worth instead having this be something like "add_assertion_that_..." and later having a "resolve_assertions"/"execute_assertions" step, or something like that? When you have "assert" here, developers have been trained by unittest to expect that the assertion is raised immediately.